base/audit_types.go

```diff
@@ -151,15 +152,25 @@ func (i AuditID) ValidateFields(f AuditFields) error {
 func mandatoryFieldsPresent(fields, mandatoryFields AuditFields, baseName string) error {
 	me := &MultiError{}
 	for k, v := range mandatoryFields {
+		if _, ok := fields[k]; !ok {
+			me = me.Append(fmt.Errorf("missing mandatory field %s", baseName+k))
+			continue
+		}
+		if !matchingTypes(v, fields[k]) {
+			me = me.Append(fmt.Errorf("field value for %s%s must be of type %T but had %T", baseName, k, v, fields[k]))
+			continue
+		}
 		// recurse if map
 		if vv, ok := v.(map[string]any); ok {
 			if pv, ok := fields[k].(map[string]any); ok {
 				me = me.Append(mandatoryFieldsPresent(pv, vv, baseName+k+"."))
 			}
 		}
-		if _, ok := fields[k]; !ok {
-			me = me.Append(fmt.Errorf("missing mandatory field %s", baseName+k))
-		}
 	}
 	return me.ErrorOrNil()
 }
+
+// matchingTypes returns true if the types of a and b are the same.
+func matchingTypes(a, b any) bool {
+	return reflect.TypeOf(a).Kind() == reflect.TypeOf(b).Kind()
```

Comment:
This is too strict for fields we rely on JSON marshalling for... Need to rethink with some spare time.

```
field value for timestamp must be of type string but had time.Time
field value for id must be of type int but had uint64
```